README.md
# Module: PushBulletNotes

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

This module will display new phone notifications onto your mirror using the PushBullet API.
This module is heavily drawn from GitHub user ronny3050's [Phone Notification Mirror](https://github.com/ronny3050/phone-notification-mirror), which stopped working a little under 9 months ago after PushBullet made a major update to their API. The general design is similar, although it has been cleaned up and adapted to fit the new API.

![](demo_images/homescreen.png)

![](demo_images/messages.png)


## API Token
You will need a [PushBullet](https://www.pushbullet.com/) app installed on your device. Once you have the app set up on your phone, you can get a PushBullet Access Token token from your Account Settings.

## Installing
Clone this repository into your __modules__ folder: `git clone https://github.com/maliciousBanjo/PushBulletNotes.git`

`cd PushBulletNotes` and run `npm install` to install the necessary dependencies.

## Dependencies
   * [PushBullet](https://www.npmjs.com/package/pushbullet)

## Development Status
Maintained, will try to tweak according to user feedback.

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
		{
			module: "PushBulletNotes",
			header: "Notifications",
			disabled: false,
			position: "bottom_left", // Or any valid MagicMirror position.
			config: {
				accessToken: "YOUR ACCESS TOKEN",
				numberOfNotifications: 5,
				displayNotificationIcon: true,
				displayMessage: true,
				displayCount: true,
				fade: true,
				maxMsgCharacters: 50,
				maxHeaderCharacters: 32
			}
		},
    ]
}
```

## Configuration options

<table width="100%">
	<thead>
		<tr>
			<th>Option</th>
			<th width="100%">Description</th>
		</tr>
	<thead>
	<tbody>
		<tr>
			<td><code>accessToken</code></td>
			<td>Your Pushbullet Access Token, accessible through your PushBullet Account Settings<br>
			</td>
		</tr>
		<tr>
			<td><code>numberOfNotifications</code></td>
			<td>Number of notifications to display at a time<br>
				<br><b>Default value:</b> <code>5</code>
				<br>If set to 0, no notification messages will be displayed.
			</td>
		</tr>
		<tr>
			<td><code>showNoticiationIcon</code></td>
			<td>Show the app icon generated by the notification<br>
				<br><b>Possible values:</b> <code>true</code> or <code>false</code>
				<br><b>Default value:</b> <code>true</code>
			</td>
		</tr>
		<tr>
			<td><code>showMessage</code></td>
			<td>Show the body of the notification<br>
				<br><b>Possible values:</b> <code>true</code> or <code>false</code>
				<br><b>Default value:</b> <code>true</code>
			</td>
		</tr>
		<tr>
			<td><code>showCount</code></td>
			<td>Show the notification count<br>
				<br><b>Possible values:</b> <code>true</code> or <code>false</code>
				<br><b>Default value:</b> <code>true</code>
			</td>
		</tr>
		<tr>
			<td><code>maxMsgCharacters</code></td>
			<td>The maximum number of characters to show from the notification body<br>
				<br><b>Default value:</b> <code>50</code>
			</td>
		</tr>
		<tr>
			<td><code>maxHeaderCharacters</code></td>
			<td>The maximum number of characters to show from the notification header<br>
				<br><b>Default value:</b> <code>32</code>
			</td>
		</tr>
</table>

CHANGELOG.md
# 
PushBulletNotes Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.1]

Added some debugging lines

## [1.1.0] - Released

Bugfix: Missing Logging

## [1.1.0] - Released

Added a new config option to limit the amount of characters in a notification header

Dead message handling

## [1.0.0] - Released

First public release

node_helper.js
/* Magic Mirror
 * Node Helper: PushBulletNotes
 *
 * By maliciousbanjo
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var PushBullet = require("pushbullet");

module.exports = NodeHelper.create({

	socketNotificationReceived: function(notification, payload) {
		if (notification === "START") {
			this.config = payload; // The payload is all of our configuration options
			console.log(notification + " received");
			this.pushBulletListener(this.config); // Start up the PushBullet listener
		}
	},

	pushBulletListener(config){
		var pusher = new PushBullet(config.accessToken); // PushBulletAPI object
		var stream = pusher.stream();
		var self = this;
		stream.connect();

		stream.on("push", function(push){ // Push is a JSON
			console.log("Push Received");
			console.log(push);
			if (push.type === "sms_changed") { // Text message
				if (push.notifications.length !== 0) {
					var message = {
						application_name: "sms",
						package_name: "sms",
						title: push.notifications[0].title,
						body: push.notifications[0].body,
						// Come back to this later, maybe make an independent window for a text profile
						// It's too small to really be discernable of anything.
						icon: (typeof(push.notifications[0].image_url) == "undefined" ? null : push.notifications[0].image_url)
					}
					console.log("SMS received, sending to mirror");
					console.log(push);
					self.sendSocketNotification("SMS_MESSAGE", message);
				}
				else {
					console.log("Dead Message");
					// Do nothing
				}
			}
			if (push.type === "mirror") { // Ordinary notification
				var notification = {
					application_name: push.application_name,
					id: push.notification_id,
					package_name: push.package_name,
					count: 1,
					icon: push.icon, // base64 data
					title: push.title,
					body: push.body
				};
				console.log("Notification received, sending to mirror");
				self.sendSocketNotification("NOTIFICATION", notification);
			}
			if (push.type === "dismissal") { // Dismissal
				var dismissal = {
					package_name: push.package_name,
					notification_id: push.notification_id
				};
				console.log("Dismissal received, sending to mirror");
				self.sendSocketNotification("DISMISSAL", dismissal);
			}
		});
	},
});

PushBulletNotes.js
/* Magic Mirror
 * Module: PushBulletNotes
 *
 * By maliciousbanjo
 * MIT Licensed.
 */

Module.register("PushBulletNotes", {
	defaults: {
		accessToken: "",
		numberOfNotifications: 5,
		showNotificationIcon: true,
		showMessage: true,
		showCount: false,
		fade: true,
		maxMsgCharacters: 50,
		maxHeaderCharacters: 32
	},

	payload: [],

	start: function() {
		console.log("PushBulletNotes module started!");
		//Flag for check if module is loaded
		this.loaded = false;
		this.sendSocketNotification("START", this.config);
		this.originalHeader = this.data.header;
	},

	getDom: function() {
		var wrapper = document.createElement("table");
		wrapper.className = "small";
		var self = this;

		if (this.config.showCount) { // Display the number of messages
			var headerRow = document.createElement("tr");
			var headerData = document.createElement("td");
			var headerDiv = document.createElement("div");

			headerDiv.className = "count";
			headerDiv.innerHTML = this.payload.length;
			headerData.appendChild(headerDiv);

			var headerRest = document.createElement("td");

			headerRest.innerHTML = this.originalHeader;
			headerRow.appendChild(headerData);
			headerRow.appendChild(headerRest);
			wrapper.appendChild(headerRow);
		}

		if (this.payload.length > 0) {
			var count = 0;

			// Only display however many notifications are specified by the config
			this.payload.slice(0, this.config.numberOfNotifications).forEach(function(o) {
				var name = (o.application_name == "sms") ? o.title : o.application_name; // If SMS, o.title is the contact name
				var title = (o.application_name == "sms") ? "" : o.title;

				// Determine if the header texts need truncating
				if (name.length + title.length > this.config.maxHeaderCharacters) {
					var remainder = (name.length + title.length) - this.config.maxHeaderCharacters;
					if (o.application_name == "sms") { // SMS Message
						name = name.substring(0, name.length - (remainder + 1)) + "...";
					}
					else {
						title = title.substring(0, title.length - (remainder + 1)) + "...";
					}
				}

				var notificationWrapper = document.createElement("tr");
				notificationWrapper.className = "normal";

				if (self.config.showNotificationIcon) {
					var iconWrapper = document.createElement("td");
					iconWrapper.className = "icon";
					var icon = document.createElement("span");
					if (o.application_name == "sms") { // Text message, display icon
						var smsPath = "/modules/PushBulletNotes/icons/sms.png";
						icon.innerHTML = "<img src=\"" + smsPath + "\" width=\"25\" >";
					}
					else {
						icon.innerHTML = "<img src=\"data:image/png;base64, " + o.icon + "\" width=\"25\" >";
					}
					iconWrapper.appendChild(icon);
					notificationWrapper.appendChild(iconWrapper);
				}

				var nameWrapper = document.createElement("td");
				nameWrapper.className = "bright";
				nameWrapper.innerHTML = name;
				notificationWrapper.appendChild(nameWrapper);

				var titleWrapper = document.createElement("td");
				titleWrapper.className = "bright";
				titleWrapper.innerHTML = title;
				notificationWrapper.appendChild(titleWrapper);
				wrapper.appendChild(notificationWrapper);

				if (self.config.showMessage) {
					var bodyWrapper = document.createElement("tr");
					var bodyContentWrapper = document.createElement("td");
					bodyContentWrapper.colSpan = "3";
					bodyContentWrapper.className = "dimmed xsmall address";
					bodyContentWrapper.innerHTML = o.body.substring(0, self.config.maxMsgCharacters);
					bodyWrapper.appendChild(bodyContentWrapper);
					wrapper.appendChild(bodyWrapper);
				}

				// Fade effect
				if (self.config.fade) {
					var startingPoint = self.payload.slice(0, self.config.numberOfNotifications).length * 0.25;
					var steps = self.payload.slice(0, self.config.numberOfNotifications).length - startingPoint;
					if (count >= startingPoint) {
						var currentStep = count - startingPoint;
						notificationWrapper.style.opacity = 1 - (1 / steps * currentStep);
					}
				}
				count++;
			});
		}
		else {
			wrapper.innerHTML = this.translate("No new notifications");
			wrapper.className = "small dimmed";
			return wrapper;
		}
		return wrapper;
	},

	getScripts: function() {
		return [];
	},

	getStyles: function () {
		return [
			"PushBulletNotes.css",
		];
	},

	cleanPayload: function(newPayload) {
		var name = newPayload.application_name;
		var title = newPayload.title;
		var self = this;
		var dupIndex = 0;
		if (this.payload.length > 0) {
			this.payload.forEach(function(m) {
				// If application_name already exists, increment notification count
				if (m.application_name === name && m.title === title) {
					self.payload.splice(dupIndex, 1);
				}
				dupIndex++;
			});
		}
		this.payload.unshift(newPayload);
	},

	removePayload: function(dismissedPayload) {
		var name = dismissedPayload.package_name;
		var self = this;
		var index = 0;

		if (this.payload.length > 0) {
			this.payload.forEach(function(m) {
				//If package_name exists in Notification list, remove notification
				if (m.package_name === name) {
					self.payload.splice(index, 1);
				}
				index++;
			});
		}
	},

	socketNotificationReceived: function (notification, payload) {
		console.log(notification);
		if(notification === "SMS_MESSAGE") {
			if (payload) {
				this.loaded = true;
				this.cleanPayload(payload);
			}
			this.updateDom();
		}
		else if (notification === "NOTIFICATION") {
			if (payload) {
				this.loaded = true;
				this.cleanPayload(payload);
			}
			this.updateDom();
		}
		else if (notification === "DISMISSAL") {
			if (payload) {
				this.loaded = true;
				this.removePayload(payload);
				this.updateDom();
			}
		}
	},
});
